psi4/driver/qcdb/intf_dftd3/dashparam.py

```diff
@@ -0,0 +1,608 @@
+            dashlevel_candidate_2 = get_dispersion_aliases()[trial_split[1]]
+            name_key = trial_split[0]
+            disp_params = dashcoeff[dashlevel_candidate_2]['definitions'][trial_split[0]]['params']
+        elif (level_hint is not None) and name_hint in dashcoeff[dashlevel_candidate_1]['definitions']:
+            # case fctl: b3lyp
+            dashlevel_candidate_2 = None
+            name_key = name_hint
+            disp_params = dashcoeff[dashlevel_candidate_1]['definitions'][name_hint]['params']
+        elif name_hint == '':
+            dashlevel_candidate_2 = None
+            name_key = None
+            disp_params = {}
+        else:
+            dashlevel_candidate_2 = None
+            raise ValidationError("""Can't guess -D correction level from ({})""".format(name_hint))
+
+    disp_params = copy.deepcopy(disp_params)
+
+    if dashlevel_candidate_1 is None and dashlevel_candidate_2 is None:
+        raise ValidationError(
+            """Can't guess -D correction level from name_hint ({}) and level_hint ({})""".format(level_hint))
+    elif dashlevel_candidate_1 is not None and dashlevel_candidate_2 is not None:
+        if dashlevel_candidate_1 != dashlevel_candidate_2:
+            raise ValidationError(
+                """Inconsistent -D correction level from name_hint ({}) and level_hint ({})""".format(
+                    dashlevel_candidate_2, dashlevel_candidate_1))
+    dashleveleff = dashlevel_candidate_1 or dashlevel_candidate_2
+
+    allowed_params = dashcoeff[dashleveleff]['default'].keys()
+
+    # << 2 >> use name_hint and/or param_tweaks to determine intended dispersion parameters
+    if name_hint is None and param_tweaks is None:
+        raise ValidationError("""Can't guess -D parameters without name_hint ({}) or param_tweaks ({})""".format(
+            name_hint, param_tweaks))
+
+    if isinstance(param_tweaks, list):
+        param_tweaks = dict(zip(allowed_params, param_tweaks))
+
+    if param_tweaks is None:
+        param_tweaks = {}
+
+    if not set(param_tweaks.keys()).issubset(allowed_params):
+        raise ValidationError('Requested keys ({}) not among allowed ({}) for dispersion level ({})'.format(
+            list(param_tweaks.keys()), list(allowed_params), dashleveleff))
+
+    disp_params.update(param_tweaks)
+
+    if disp_params.keys() != allowed_params:
+        raise ValidationError('Requested keys ({}) insufficient ({}) for dispersion level ({})'.format(
+            list(param_tweaks.keys()), list(allowed_params), dashleveleff))
+
+    # << 3 >> use final dashlevel and disp_params to determine if a defined "fctl-disp" label exists
+    # * if/elif chooses right label when some fctls have identical param sets
+    if ((name_hint is not None) and (dashcoeff_supplement is not None)
+            and (name_key in dashcoeff_supplement[dashleveleff]['definitions'])
+            and (disp_params == dashcoeff_supplement[dashleveleff]['definitions'][name_key]['params'])):
```

Comment:
Is this line here why you needed the `deepcopy` above - so that you don't overwrite `supplement` when you modify `disp_params`?

Reply:
Yes. The `disp_params` on rhs of 535 are still pointing to `dashcoeff` or `dashcoeff_supplement` objects. Because want to manipulate here and in 564 `disp_params.update(param_tweaks)`, best to grab an independent copy.